File: Python/0287_FindTheDuplicateNumber.py
class Solution:
    # Solution 1 - Binary Search
    def findDuplicate(self, nums: List[int]) -> int:
        low, high = 0, len(nums) - 1
        while low <= high:
            count = 0
            mid = low + (high - low) // 2
            for num in nums:
                count += 1 if num <= mid else 0
            if count > mid:
                high = mid - 1
            else:
                low = mid + 1

        return low

File: C++/0287_FindTheDuplicateNumber.cpp
class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int low = 0, high = nums.size() - 1;
        while (low <= high) {
            int count = 0;
            int mid = low + (high - low) / 2;
            for (int i = 0; i < nums.size(); ++i) 
                count += nums[i] <= mid ? 1 : 0;
            if (count > mid) 
                high = mid - 1;
            else
                low = mid + 1;
        }
        return low;
    }
};